src/Features/CSharp/Portable/AddParameter/CSharpAddParameterCodeFixProvider.cs

```diff
@@ -39,5 +43,49 @@ protected override ImmutableArray<string> TooManyArgumentsDiagnosticIds
 
         protected override ImmutableArray<string> CannotConvertDiagnosticIds
             => GenerateConstructorDiagnosticIds.CannotConvertDiagnosticIds;
+
+        protected override async Task<bool> HandleLanguageSpecificNodesAsync(SyntaxNode initialNode, SyntaxNode node, CodeFixContext context)
+        {
+            if (node is ConstructorInitializerSyntax constructorInitializer)
+            {
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+                var constructorDeclaration = constructorInitializer.Parent;
+                if (semanticModel.GetDeclaredSymbol(constructorDeclaration, cancellationToken) is IMethodSymbol constructorSymbol)
+                {
+                    var type = constructorSymbol.ContainingType;
+                    if (constructorInitializer.IsKind(SyntaxKind.BaseConstructorInitializer))
+                    {
+                        // Search for fixable constructors in the base class.
+                        type = type?.BaseType;
+                    }
+
+                    if (type != null && type.IsFromSource())
+                    {
+                        var methodCandidates = type.InstanceConstructors;
+                        if (constructorInitializer.IsKind(SyntaxKind.ThisConstructorInitializer))
+                        {
+                            // Exclude the constructor with the diagnostic from the list of candidates
+                            methodCandidates = methodCandidates.Remove(constructorSymbol);
+                        }
+
+                        var arguments = constructorInitializer.ArgumentList.Arguments;
+                        var argumentOpt = TryGetRelevantArgument(initialNode, node, context.Diagnostics.First());
+
+                        var insertionData = GetArgumentInsertPositionForMethodCandidates(
+                            argumentOpt, semanticModel, syntaxFacts, arguments, methodCandidates);
+
+                        RegisterFixForMethodOverloads(context, arguments, insertionData);
```

Comment:
nit (And you def don't ahve to do anything here), but i don't like the imperative nature of all the code we have here.  once thing i think i might like more would be if these helper methods (like HandleLanguageSpecificNodesAsync) returned a type representing "arguments/insertionData".  Then it would be the *callers* responsibility to see if this was returned, and it would take care of registering the fix. 

Effectively, i don't like that this subclass has any ability to do this or see these methods.  

Does that make sense?   If you disagree, then i'm not going to press the point.  But it just feels a bit... cleaner to me.  Thanks! :)

Reply:
I agree that the `RegisterCodeFixesAsync` method could benefit from some refactorings.
https://github.com/MaStr11/roslyn/blob/472f16c1c4266f075ad7ec8ad08e98e931ab1af9/src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs#L54-L72

There is also some duplication of code with `var argumentOpt = TryGetRelevantArgument(initialNode, node, diagnostic);`, but I couldn't find a good solution yesterday. Let me try out some refactorings later (I'm on vacation and on the way to the beach right now).

There are also the indexers for [C#](https://github.com/dotnet/roslyn/blob/8e22089fbb59f2d68c4a7b6b09f20fdd8de0e36e/src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs#L2390) and [VB](https://github.com/dotnet/roslyn/blob/8e22089fbb59f2d68c4a7b6b09f20fdd8de0e36e/src/EditorFeatures/VisualBasicTest/AddParameter/AddParameterTests.vb#L1022) that probably need to be resolved in a language-specific way, so the refactoring might take those into consideration too.

Reply:
I refactored `RegisterCodeFixesAsync` in 1a51cb7. I got rid of all the code duplication and think the code is more easy to maintain now. The naming of the methods (`GetDataForFix_ ...`) is poor and the extra `RegisterFixData` type might be overkill, but I think the implementation comes close to what you intended.

I also took a brief look into the indexer problem mentioned earlier. It seems it requires indeed a language-specific solution as the syntax trees are very different (I think there are more problems than just that, but that would be part of a follow-up PR if I/you/anyone ever wants to add support for indexers).